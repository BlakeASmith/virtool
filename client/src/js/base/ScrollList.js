import React from "react";
import { map } from "lodash-es";
import { LoadingPlaceholder } from "./index";

export class ScrollList extends React.Component {

    componentDidMount () {
        window.addEventListener("scroll", this.onScroll, false);
    }

    componentWillUnMount () {
        window.removeEventListener("scroll", this.onScroll, false);
    }

    onScroll = () => {
        // Scroll bar reaches the bottom of page when ratio = 1
        // Want to load available page when scroll bar nears the bottom
        const totalHeight = document.body.scrollHeight;
        const windowHeight = window.innerHeight;
        const distScrolled = window.scrollY;
        const ratio = ((windowHeight + distScrolled) / totalHeight).toFixed(1);

        if (this.props.list.length
            && this.props.hasNextPage
            && !this.props.isNextPageLoading
            && (ratio > 0.8)
        ) {
            this.props.loadNextPage(this.props.page + 1);
        }
    };

    render () {
        const { list, rowRenderer, isNextPageLoading } = this.props;

        return (
            <div>
                {map(list, (item, index) => rowRenderer(index))}
                {isNextPageLoading ? <LoadingPlaceholder margin="20px" /> : null}
            </div>
        );
    }
}
