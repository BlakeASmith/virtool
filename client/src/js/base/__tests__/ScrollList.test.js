import { ScrollList, calculateScrollRatio, getEventTarget } from "../ScrollList";
import { LoadingPlaceholder } from "../index";

describe("<ScrollList />", () => {
    let props;
    let state;

    beforeEach(() => {
        props = {
            isElement: true,
            document: [{ foo: "bar" }],
            page: 1,
            pageCount: 2,
            onLoadNextPage: jest.fn(),
            style: "baz"
        };
        state = {};
    });

    it("should render", () => {
        const wrapper = shallow(<ScrollList {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});
