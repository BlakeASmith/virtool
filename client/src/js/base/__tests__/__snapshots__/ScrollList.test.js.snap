// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrollList /> should render when [noContainer=false] 1`] = `
<ScrollList
  document={
    Array [
      Object {
        "foo": "bar",
      },
    ]
  }
  isElement={true}
  noContainer={false}
  onLoadNextPage={[MockFunction]}
  page={1}
  pageCount={2}
>
  <div />
</ScrollList>
`;

exports[`<ScrollList /> should return LoadingPlaceholder when [documents=null] and [page<pageCount] 1`] = `
<ScrollList
  document={
    Array [
      Object {
        "foo": "bar",
      },
    ]
  }
  documents={null}
  isElement={true}
  onLoadNextPage={[MockFunction]}
  page={1}
  pageCount={2}
>
  <div>
    <LoadingPlaceholder
      margin="20px"
    >
      <div
        className="text-center"
        style={
          Object {
            "marginTop": "20px",
          }
        }
      >
        <Loader>
          <Loader__StyledLoader
            color="#3c8786"
            size="22px"
          >
            <div
              className="Loader__StyledLoader-sc-338y7h-0 cRPClh"
              color="#3c8786"
              size="22px"
            >
              <div />
            </div>
          </Loader__StyledLoader>
        </Loader>
      </div>
    </LoadingPlaceholder>
  </div>
</ScrollList>
`;

exports[`<ScrollList /> should return React.Fragment when [noContainer=true] 1`] = `
<ScrollList
  document={
    Array [
      Object {
        "foo": "bar",
      },
    ]
  }
  isElement={true}
  onLoadNextPage={[MockFunction]}
  page={1}
  pageCount={2}
>
  <div />
</ScrollList>
`;
